Data: Await resolver implementing isFulfilled

johnhooks opened the pull request:
## What?

Add dispatch of `finishResolution` when a resolver's `isFulfilled` callback is true in the `@wordpress/data` package.

Fixes #70805.

## Why?

A deadlock issue was discovered in the `@wordpress/data` package when using `resolveSelect` on a selector/resolver pair implementing `isFulfilled`.

This [failed test run](https://github.com/WordPress/gutenberg/actions/runs/16402291352/job/46343436719?pr=70806#step:6:220) from before adding the fix demonstrates the issue.  

## How?

The concept is to dispatch `finishResolution` to the metadata store when `resolver.isFulfilled` is `true`. This adds the correct state to the store, allowing resolution of the promises in `mapResolveSelector` and `mapSuspendSelector`.

johnhooks committed:
Await resolver implementing isFulfilled

Add a test to the `@wordpress/data` package to demostrate the deadlock of
awaiting a resolver implementing `isFulfilled`.

See #70805.

johnhooks committed:
Fix test resolvedState

johnhooks committed:
Add more realistic isFulfilled callback

johnhooks committed:
Add fulfillSelector fix

johnhooks committed:
Colocate tests

johnhooks commented:
@Mamaduka I've updated the branch, colocated the tests, and added a few more after reviewing the other tests in `packages/data/src/test/registry.js`.

jsnajdr commented on packages/data/src/redux-store/index.js:
```diff
@@ -700,19 +700,29 @@ function mapSelectorWithResolver(
+				store.dispatch(
+					metadataActions.finishResolution( selectorName, args )
+				);
+			}, 0 );
```
This could be neatly integrated into the main `setTimeout` callback several lines below. Just don't run the `fulfill` function when `isFulfilled` returns `true`:
```js
const isFulfilled =
  typeof resolver.isFulfilled === 'function' &&
  resolver.isFulfilled( state, ...args );
if ( ! isFulfilled ) {
  const action = resolver.fulfill( ...args );
  if ( action ) {
    await store.dispatch( action );
  }
}
```

It's the same thing, without code duplication, with just one difference: both `startResolution` and `finishResolution` are called, right after each other without any waiting. But that's fine, it's even welcome, because outside observers will see the same sequence of updates, no matter how the selector/resolver is internally implemented.

Another way to fix this is in the `mapResolveSelector` and `mapSuspendSelector` functions. These functions could perform the `isFulfilled` check and return a result right away. In that case, the `startResolution`/`finishResolution` actions would never be called when `isFulfilled` is `true`. But it's probably a worse solution than the first one, again for consistency reasons: sometimes, for the same selector, `start`/`finishResolution` would be called, sometimes not.

johnhooks replied:
I agree with calling `startResolution`/`finishResolution` and being consistent. I'll wrap the call to `isFulfilled` into the main `setTimeout`. Thank you for the review.

johnhooks committed:
Move isFulfilled check inside setTimeout callback

Reduces duplication by handling isFulfilled in the same code path as fulfill.
Both paths now mark the resolver as running first, and isFulfilled is now
wrapped in try/catch for error handling.

johnhooks committed:
Add isFulfilled throw test

johnhooks commented on packages/data/src/test/registry.js:
```diff
@@ -358,6 +358,150 @@ describe( 'createRegistry', () => {
+
+			const promise = registry.resolveSelect( 'demo' ).getData();
+			jest.runAllTimers();
+			await expect( promise ).rejects.toThrow( 'isFulfilled error' );
```
@jsnajdr this is odd. The error is caught and the state of the resolver transitions to failed, but Jest fails the test without this line. I'm still researching the issue.

johnhooks replied:
I've walked through with the debugger, wow there is a lot going on for a single resolution. Though after running `jest.runAllTimers`, `promise` has a rejected status and `Error: isFulfilled error`, even though it was caught in the try/catch and included in the `FAIL_RESOLUTION` action.

johnhooks replied:
OH I got it, [`mapResolveSelector`](https://github.com/WordPress/gutenberg/blob/d3734f5e4839d0cf5ada56fedb253fae26c55116/packages/data/src/redux-store/index.js#L586) rejects it. So the test is correct, if this is the desired behavior.

Should the resolver resolution fail if `isFulfilled` throws an error?

jsnajdr replied:
Yes, `resolveSelect` detects the error and returns it.

If `isFulfilled` throws, that is certainly suprising and unexpected and the error should be surfaced. So yes, the resolution should fail.

jsnajdr commented on packages/data/src/redux-store/index.js:
```diff
@@ -700,22 +700,15 @@ function mapSelectorWithResolver(
 			return;
 		}
 
+		const state = store.getState();
```
This could be inlined into the `isFulfilled` call, we don't use the `state` anywhere else do we?
```js
resolver.isFulfilled( store.getState(), ...args )
```

jsnajdr commented on packages/data/src/test/registry.js:
```diff
@@ -358,6 +358,150 @@ describe( 'createRegistry', () => {
+			} );
+
+			const promise1 = registry.resolveSelect( 'demo' ).getPage( 1 );
+			jest.runAllTimers();
```
For the `resolveSelect` tests, we would be better off if we used `jest.useRealTimers()`. Then you don't need to run timers manually.

It will happen automatically if you move these tests to `redux-store/test/index.js`, where all other `resolveSelect` tests are. This `registry.js` suite has `jest.useFakeTimers()` at the start.

johnhooks committed:
Colocate with other resolveSelect tests

jsnajdr reviewed (approved):
Very nice, thanks for working on this and addressing all feedback :+1:

johnhooks commented:
@jsnajdr absolutely, I appreciate you taking the time to help me through it.

Also shout out to @TimothyBJacobs for suggesting I create the PR.
